python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -0,0 +1,233 @@
+
+                    # use scatter to sync across trainers about the p2p tensor size
+                    # Note: If we have GPU nccl support, we can use all_to_all to
+                    # sync information here
+                    gather_list = list(th.empty([self._world_size],
+                                                dtype=th.int64).chunk(self._world_size))
+                    alltoall_cpu(self._rank, self._world_size, gather_list, idx_split_size)
+                    # use cpu until we have GPU alltoallv
+                    idx_gather_list = [th.empty((int(num_emb),),
+                                                dtype=idics.dtype) for num_emb in gather_list]
+                    alltoallv_cpu(self._rank, self._world_size, idx_gather_list, idics_list)
+                    local_indics[name] = idx_gather_list
+                    grad_gather_list = [th.empty((int(num_emb), grads.shape[1]),
+                                                 dtype=grads.dtype) for num_emb in gather_list]
+                    alltoallv_cpu(self._rank, self._world_size, grad_gather_list, grad_list)
+                    local_grads[name] = grad_gather_list
+                else:
+                    local_indics[name] = [idics]
+                    local_grads[name] = [grads]
+
+            if self._clean_grad:
+                # clean gradient track
+                for emb in self._params:
+                    emb.reset_trace()
+                self._clean_grad = False
+
+            # do local update
+            for emb in self._params:
+                name = emb._tensor.name
+
+                idx = th.cat(local_indics[name], dim=0)
+                grad = th.cat(local_grads[name], dim=0)
+                self.update(idx.to(device, non_blocking=True),
+                            grad.to(device, non_blocking=True), emb)
+
+        # synchronized gradient update
+        if self._world_size > 1:
+            th.distributed.barrier()
+
+    @abstractmethod
+    def update(self, idx, grad, emb):
+        """ Update embeddings in a sparse manner
+        Sparse embeddings are updated in mini batches. we maintains gradient states for
+        each embedding so they can be updated separately.
+
+        Parameters
+        ----------
+        idx : tensor
+            Index of the embeddings to be updated.
+        grad : tensor
+            Gradient of each embedding.
+        emb : dgl.distributed.NodeEmbedding
+            Sparse node embedding to update.
+        """
+
+    def zero_grad(self):
+        """clean grad cache
+        """
+        self._clean_grad = True
+
+def initializer(shape, dtype):
+    """ Sparse optimizer state initializer
+
+    Parameters
+    ----------
+    shape : tuple of ints
+        The shape of the state tensor
+    dtype : torch dtype
+        The data type of the state tensor
+    """
+    arr = th.zeros(shape, dtype=dtype)
+    return arr
+
+class SparseAdagrad(DistSparseGradOptimizer):
```

Comment:
this is synchronized update. we should keep the async update so that we can make a comparison between them.